parsl/executors/workqueue/executor.py

```diff
@@ -729,14 +729,25 @@ def _collect_work_queue_results(self):
                 with self.tasks_lock:
                     future = self.tasks.pop(task_report.id)
                 logger.debug("Updating Future for executor task {}".format(task_report.id))
+                # If result_received, then there's a result file. The object inside the file
+                # may be a valid result or an exception caused within the function invocation.
+                # Otherwise there's no result file, implying errors from WorkQueue.
                 if task_report.result_received:
-                    future.set_result(task_report.result)
+                    try:
+                        with open(task_report.result_file, 'rb') as f_in:
+                            result = deserialize(f_in.read())
+                    except Exception as e:
+                        logger.debug(f'Cannot load result from result file {task_report.result_file}. Exception: {e}')
+                        future.set_exception(WorkQueueTaskFailure('Cannot load result from result file', None))
+                    else:
+                        if isinstance(result, Exception):
+                            future.set_exception(result)
+                        else:
+                            future.set_result(result)
                 else:
                     # If there are no results, then the task failed according to one of
                     # work queue modes, such as resource exhaustion.
-                    ex = WorkQueueTaskFailure(task_report.reason, task_report.result)
-                    if task_report.result is not None:
-                        ex.__cause__ = task_report.result
+                    ex = WorkQueueTaskFailure(task_report.reason, Exception(task_report.reason))
```

Comment:
I think task_report.result is an exception object that has its own stack trace and more detailed reason for failure:

I feel like that exception object should be chained on using `__cause__` to give more detailed stack traces that preserve the information in the original exception, based on the principle that this code (like most code) is not in the right place to discard exception information.

Likewise for a bunch of other places where there's a task failure exception in an exception context (eg line 741, and the equivalents in taskvine)

Reply:
In the line you referred to there's no result file transferred back meaning that it's an internal WQ error. The exception chaining happens in line 743 where a result file comes back and it contains an exception. The executor sets the original exception then. I'll wrap the exception into a WQ Task Failure to make it clearer.

Reply:
right, but if there is *ever* an exception *anywhere*, I don't want it ever discarded apart from at the very outer fringe of parsl

Reply:
I think the new change adds function chaining whenever it is possible for WQ. Am I understanding you correctly @benclifford ?

Reply:
@tphung3 yes looks good for Work Queue - maybe Task Vine should have the same behaviour, though? This bit of the code is basically the same in both implementations. Other than that, this is ready to merge I think.